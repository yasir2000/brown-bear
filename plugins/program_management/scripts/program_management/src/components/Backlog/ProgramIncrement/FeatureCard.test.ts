/**
 * Copyright (c) Enalean, 2021 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { Wrapper } from "@vue/test-utils";
import { shallowMount } from "@vue/test-utils";
import FeatureCard from "./FeatureCard.vue";
import { createProgramManagementLocalVue } from "../../../helpers/local-vue-for-test";
import { createStoreMock } from "@tuleap/core/scripts/vue-components/store-wrapper-jest";
import type { ProgramIncrement } from "../../../helpers/ProgramIncrement/program-increment-retriever";
import FeatureCardBacklogItems from "./FeatureCardBacklogItems.vue";
import type { Feature } from "../../../type";
import type { ConfigurationState } from "../../../store/configuration";

describe("FeatureCard", () => {
    const getWrapper = async (
        feature?: Partial<Feature>,
        configuration?: Partial<ConfigurationState>,
        user_can_plan = true,
        ongoing_move_elements_id: number[] = []
    ): Promise<Wrapper<FeatureCard>> => {
        const defaulted_feature = {
            id: 100,
            title: "My artifact",
            tracker: {
                label: "bug",
                color_name: "lake_placid_blue",
            },
            is_open: true,
            background_color: "",
            has_user_story_planned: false,
            has_user_story_linked: false,
            ...feature,
        };

        const defaulted_configuration = {
            accessibility: false,
            has_plan_permissions: true,
            ...configuration,
        };

        const component_options = {
            propsData: {
                feature: defaulted_feature,
                program_increment: {
                    user_can_plan,
                } as ProgramIncrement,
            },
            localVue: await createProgramManagementLocalVue(),
            mocks: {
                $store: createStoreMock({
                    state: {
                        configuration: defaulted_configuration,
                        ongoing_move_elements_id,
                    },
                }),
            },
        };
        return shallowMount(FeatureCard, component_options);
    };

    it("Displays a draggable card with accessibility pattern", async () => {
        const wrapper = await getWrapper(
            { background_color: "peggy_pink_text" },
            { accessibility: true }
        );
        expect(wrapper.element).toMatchSnapshot();
    });

    it(`Adds a closed class to the card`, async () => {
        const wrapper = await getWrapper({ is_open: false });
        const card = wrapper.find("[data-test=feature-card]");
        expect(card.classes()).toContain("element-card-closed");
    });

    it("Displays a not draggable card without accessibility pattern", async () => {
        const wrapper = await getWrapper({}, { has_plan_permissions: false });
        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a not draggable card when user can not plan/unplan features", async () => {
        const wrapper = await getWrapper({ has_user_story_planned: true }, {}, false);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a draggable card with items backlog container", async () => {
        const wrapper = await getWrapper(
            { has_user_story_planned: true, has_user_story_linked: true },
            {},
            false
        );
        expect(wrapper.findComponent(FeatureCardBacklogItems).exists()).toBeTruthy();
    });
});
