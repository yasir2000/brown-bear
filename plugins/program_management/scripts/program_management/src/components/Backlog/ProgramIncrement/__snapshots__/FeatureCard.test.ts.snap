// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureCard Displays a draggable card with accessibility pattern 1`] = `
<div
  class="element-backlog-item"
  data-element-id="100"
  draggable="true"
>
  <div
    class="element-card-container"
    data-tlp-tooltip=""
  >
    <div
      class="element-card element-card-lake_placid_blue element-card-background-peggy_pink_text element-card-with-accessibility"
      data-test="feature-card"
    >
      <div
        class="element-card-content"
      >
        <div
          class="element-card-xref-label"
        >
          <a
            class="element-card-xref element-card-xref-lake_placid_blue"
            data-not-drag-handle="true"
            href="/plugins/tracker/?aid=100"
          >
            
                        
                    
          </a>
           
          <span
            class="element-card-label"
          >
            My artifact
          </span>
        </div>
      </div>
       
      <div
        class="element-card-accessibility"
      />
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`FeatureCard Displays a not draggable card when user can not plan/unplan features 1`] = `
<div
  class="element-backlog-item"
  data-element-id="100"
  draggable="false"
>
  <div
    class="element-card-container tlp-tooltip tlp-tooltip-left"
    data-tlp-tooltip="The feature is not plannable, user does not have permission to update artifact or field link."
  >
    <div
      class="element-card element-card-lake_placid_blue"
      data-test="feature-card"
    >
      <div
        class="element-card-content"
      >
        <div
          class="element-card-xref-label"
        >
          <a
            class="element-card-xref element-card-xref-lake_placid_blue"
            data-not-drag-handle="true"
            href="/plugins/tracker/?aid=100"
          >
            
                        
                    
          </a>
           
          <span
            class="element-card-label"
          >
            My artifact
          </span>
        </div>
      </div>
       
      <!---->
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`FeatureCard Displays a not draggable card without accessibility pattern 1`] = `
<div
  class="element-backlog-item"
  data-element-id="100"
  draggable="false"
>
  <div
    class="element-card-container tlp-tooltip tlp-tooltip-left"
    data-tlp-tooltip="You cannot plan items"
  >
    <div
      class="element-card element-card-lake_placid_blue"
      data-test="feature-card"
    >
      <div
        class="element-card-content"
      >
        <div
          class="element-card-xref-label"
        >
          <a
            class="element-card-xref element-card-xref-lake_placid_blue"
            data-not-drag-handle="true"
            href="/plugins/tracker/?aid=100"
          >
            
                        
                    
          </a>
           
          <span
            class="element-card-label"
          >
            My artifact
          </span>
        </div>
      </div>
       
      <!---->
    </div>
  </div>
   
  <!---->
</div>
`;
